spring-boot/proxy/src/main/java/eu/coatrack/proxy/security/LocalApiKeyManager.java

```diff
@@ -0,0 +1,122 @@
+ * limitations under the License.
+ * #L%
+ */
+
+import eu.coatrack.api.ApiKey;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.scheduling.annotation.EnableAsync;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.PostConstruct;
+import java.net.ConnectException;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Locally stores, periodically refreshes and provides API keys and their
+ * associated service APIs belonging to the gateway.
+ *
+ * @author Christoph Baier
+ */
+
+@EnableAsync
+@Service
+public class LocalApiKeyManager {
+
+    private static final Logger log = LoggerFactory.getLogger(LocalApiKeyManager.class);
+
+    private List<ApiKey> localApiKeyList = new ArrayList<>();
+    private LocalDateTime latestLocalApiKeyListUpdate = LocalDateTime.now();
+
+    private final ApiKeyFetcher apiKeyFetcher;
+    private final long numberOfMinutesTheGatewayShallWorkWithoutConnectionToAdmin;
+
+    public LocalApiKeyManager(
+            ApiKeyFetcher apiKeyFetcher,
+            @Value("${number-of-minutes-the-gateway-works-without-connection-to-admin}") long minutes) {
+        this.apiKeyFetcher = apiKeyFetcher;
+        this.numberOfMinutesTheGatewayShallWorkWithoutConnectionToAdmin = minutes;
+    }
+
+    public ApiKey findApiKeyFromLocalApiKeyList(String apiKeyValue) {
+        log.debug("Trying to find the service API associated to the API key with the value {} from the local list.",
+                apiKeyValue);
+
+        if (apiKeyValue == null) {
+            log.info("The passed API key value is null and can therefore not be checked for validity. " +
+                    "It is therefore rejected.");
+            return null;
+        }
+        Optional<ApiKey> optionalApiKey = localApiKeyList.stream().filter(
+                apiKeyFromLocalList -> apiKeyFromLocalList.getKeyValue().equals(apiKeyValue)).findFirst();
+        return extractOptionalApiKeyAndLogPossibleErrors(optionalApiKey, apiKeyValue);
+    }
+
+    private ApiKey extractOptionalApiKeyAndLogPossibleErrors(Optional<ApiKey> optionalApiKey, String apiKeyValue) {
+        if (optionalApiKey.isPresent()) {
+            log.debug("The API key with the value {} is found in the local API key list.", apiKeyValue);
+            return optionalApiKey.get();
+        } else {
+            log.info("The API key with the value {} can not be found within the local API key list " +
+                    "and is therefore rejected.", apiKeyValue);
+            return null;
+        }
+    }
+
+    public boolean wasLatestUpdateOfLocalApiKeyListWithinDeadline(ApiKey apiKey) {
+        LocalDateTime deadline = latestLocalApiKeyListUpdate.plusMinutes(
```

Comment:
This "deadline" needs to be calculated only once each time the local api key list is updated. It does not change until the next update, therefore does not make sense to calculate it on every call to this method.

Reply:
Solved.